clang/docs/HLSL/RootSignature.rst

```diff
@@ -0,0 +1,210 @@
+VersionedRootSignatureDesc is a struct that contains a RootSignatureVersion 
+and a RootSignatureDesc.
+
+.. code-block:: c++
+    struct DescriptorRange {
+    DescriptorRangeType RangeType;
+    uint32_t NumDescriptors = 1;
+    uint32_t BaseShaderRegister;
+    uint32_t RegisterSpace = 0;
+    DescriptorRangeFlags Flags = DescriptorRangeFlags::None;
+    uint32_t OffsetInDescriptorsFromTableStart = DescriptorRangeOffsetAppend;
+    };
+
+    struct RootDescriptorTable {
+    std::vector<DescriptorRange> DescriptorRanges;
+    };
+    struct RootConstants {
+    uint32_t ShaderRegister;
+    uint32_t RegisterSpace = 0;
+    uint32_t Num32BitValues;
+    };
+
+    struct RootDescriptor {
+    uint32_t ShaderRegister;
+    uint32_t RegisterSpace = 0;
+    RootDescriptorFlags Flags = RootDescriptorFlags::None;
+    };
+    struct RootParameter {
+    RootParameterType ParameterType;
+    std::variant<RootDescriptorTable, RootConstants, RootDescriptor>
+        Parameter;
+    ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct StaticSamplerDesc {
+    Filter Filter = Filter::ANISOTROPIC;
+    TextureAddressMode AddressU = TextureAddressMode::Wrap;
+    TextureAddressMode AddressV = TextureAddressMode::Wrap;
+    TextureAddressMode AddressW = TextureAddressMode::Wrap;
+    float MipLODBias = 0.f;
+    uint32_t MaxAnisotropy = 16;
+    ComparisonFunc ComparisonFunc = ComparisonFunc::LessEqual;
+    StaticBorderColor BorderColor = StaticBorderColor::OpaqueWhite;
+    float MinLOD = 0.f;
+    float MaxLOD = MaxLOD;
+    uint32_t ShaderRegister;
+    uint32_t RegisterSpace = 0;
+    ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct RootSignatureDesc {
+    std::vector<RootParameter> Parameters;
+    std::vector<StaticSamplerDesc> StaticSamplers;
+    RootSignatureFlags Flags;
+    };
+
+    struct VersionedRootSignatureDesc {
+    RootSignatureVersion Version;
+    RootSignatureDesc Desc;
+    };
+
+Things like DescriptorRangeType and RootDescriptorFlags will be enums.
+
+After parsing, the VersionedRootSignatureDesc will be translated into a 
+constant global variable in the clang AST and save to the 
+HLSLRootSignatureAttr. 
+
+For case compile to a standalone root signature blob, the global variable will 
+be saved in the ASTContext.
+
+The global variable in AST will have a struct type that represents the root signature
+layout and a initializer that contains the values like space and 
+numDescriptors of the root signature.
+
+In clang code generation, the global variable in AST will be translated into a global 
+variable with cosntant initializer in LLVM IR. 
+
+CGHLSLRuntime will generate metadata to link the global variable as root 
+signature for given entry function or just nullptr for the standalone root 
+signature blob case. 
+
+In LLVM DirectX backend, the global variable will be serialized and save into the root
+signature part of dx container when emit DXIL.
```

Comment:
Does clang actually use the root signature for anything?

Reply:
It will be used in validation.
Added to the end of the doc.